tortoisebeacon/handlers.go

```diff
@@ -0,0 +1,213 @@
+			log.Uint64("epoch_id", uint64(m.Epoch())),
+			log.Int("type", int(mt)))
+
+		return nil
+
+	default:
+		tb.Log.With().Info("Received ProposalMessage of unknown type",
+			log.Uint64("epoch_id", uint64(m.Epoch())),
+			log.String("message", m.String()))
+
+		return ErrUnknownMessageType
+	}
+}
+
+// HandleSerializedVotingMessage defines method to handle Tortoise Beacon proposal Messages from gossip.
+func (tb *TortoiseBeacon) HandleSerializedVotingMessage(data service.GossipMessage, sync service.Fetcher) {
+	from := data.Sender()
+
+	tb.Log.With().Info("New voting message",
+		log.String("from", from.String()))
+
+	var m VotingMessage
+	if err := types.BytesToInterface(data.Bytes(), &m); err != nil {
+		tb.Log.With().Error("Received invalid voting message",
+			log.String("message", string(data.Bytes())),
+			log.Err(err))
+
+		return
+	}
+
+	if err := tb.handleVotingMessage(from, m); err != nil {
+		tb.Log.With().Error("Failed to handle voting message",
+			log.String("message", m.String()),
+			log.Err(err))
+
+		return
+	}
+
+	data.ReportValidation(TBVotingProtocol)
+}
+
+func (tb *TortoiseBeacon) handleVotingMessage(from p2pcrypto.PublicKey, message VotingMessage) error {
+	messageEpoch := message.Epoch()
+	messageRound := message.Round()
+
+	mt := tb.classifyMessage(message, messageEpoch)
+	switch mt {
+	case TimelyMessage:
+		tb.Log.With().Info("Received timely VotingMessage, counting it",
+			log.Uint64("epoch_id", uint64(message.Epoch())),
+			log.Uint64("round", uint64(messageRound)),
+			log.String("message", message.String()))
+
+		thisRound := epochRoundPair{
+			EpochID: messageEpoch,
+			Round:   messageRound,
+		}
+
+		tb.votesMu.Lock()
+		defer tb.votesMu.Unlock()
+
+		if _, ok := tb.incomingVotes[thisRound]; !ok {
+			tb.incomingVotes[thisRound] = make(votesPerPK)
+		}
+
+		votesFor := make(hashSet)
+		votesAgainst := make(hashSet)
+
+		for _, vote := range message.For() {
+			votesFor[vote] = struct{}{}
+		}
+
+		for _, vote := range message.Against() {
+			votesAgainst[vote] = struct{}{}
+		}
+
+		tb.incomingVotes[thisRound][from] = votesSetPair{
+			VotesFor:     votesFor,
+			VotesAgainst: votesAgainst,
+		}
+
+		return nil
+
+	case DelayedMessage, LateMessage:
+		tb.Log.With().Info(fmt.Sprintf("Received %v VotingMessage, ignoring it", mt.String()),
+			log.Uint64("epoch_id", uint64(message.Epoch())),
+			log.Uint64("round", uint64(messageRound)),
+			log.String("message", message.String()))
+
+		return nil
+
+	default:
+		tb.Log.With().Info("Received VotingMessage of unknown type",
+			log.Uint64("epoch_id", uint64(message.Epoch())),
+			log.Uint64("round", uint64(messageRound)),
+			log.Int("type", int(mt)))
+
+		return ErrUnknownMessageType
+	}
+}
+
+func (tb *TortoiseBeacon) classifyMessage(m message, epoch types.EpochID) MessageType {
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	messageRound := types.RoundID(1)
+	if vm, ok := m.(VotingMessage); ok {
```

Comment:
if not ok - log error

Reply:
`classifyMessage` can be used for both `ProposalMessage` and `VotingMessage`. 

For `ProposalMessage`, `messageRound == 1`. For `VotingMessage`, `messageRound` is extracted from message.